### PR TITLE
TeamCity Module

### PR DESCRIPTION
**Pull request by cwwalb**
**Feature request:**  TeamCity Server on ECS Fargate #12

## Summary
TeamCity server creation with mounted EFS and Aurora Serverless V2 PostgreSQL configuration. Load balancing, access logging, and health checks configured. Allows configuration of database resources with connection strings as well as providing self managed Amazon EFS file system and access points.

### Changes

> Initial module creation with VPC and docker image running TeamCity server with mounted EFS for data and logs. Aurora Serverless V2 running PostgreSQL engine backing the TeamCity server. Resource creation is configurable, and checkov requirements are addressed.

### User experience

> Using this module, a user will be able to deploy TeamCity Server hosted on an Amazon ECS container managed by AWS Fargate. The TeamCity Server will have an attached Amazon EFS file system and Amazon Aurora Serverless v2 PostgreSQL database. The user will have the option to create these resources by deploying this module from the simple example, or configure this module to use their own application load balancer, ECS cluster, database via a database connection string, or use their own managed Amazon EFS file system.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [ ] I have performed a self-review of this change
* [ ] Changes have been tested
* [ ] Changes are documented

<details>
<summary>Is this a breaking change?</summary>

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created might not be successful.

### CODE REVIEW

File: modules/teamcity/examples/simple/main.tf [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module "teamcity" {
+  source              = "../../"
+  vpc_id              = aws_vpc.teamcity_vpc.id
+  service_subnets     = aws_subnet.private_subnets[*].id
+  alb_subnets         = aws_subnet.public_subnets[*].id
+  alb_certificate_arn = aws_acm_certificate.teamcity.arn
+
```
Comment (henrykie) [inline modules/teamcity/examples/simple/main.tf:7]: nit: remove blank line

File: modules/teamcity/examples/simple/versions.tf [hunk 2]
```diff
@@ -0,0 +1,10 @@
+terraform {
+  required_version = ">= 1.0"
```
Comment (henrykie) [inline modules/teamcity/examples/simple/versions.tf:2]: I'd like to see us bump this to `required_version = ">= 1.9"` and see if it still deploys.

File: modules/teamcity/examples/simple/versions.tf [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "5.82.2"
```
Comment (henrykie) [inline modules/teamcity/examples/simple/versions.tf:7]: This also needs to be bumped to latest. I believe its `5.85.0` now. Dependabot will keep this up to date going forward.

File: modules/teamcity/local.tf [hunk 4]
```diff
@@ -0,0 +1,9 @@
+locals {
+  name_prefix = "teamcity"
+  tags = {
+
+  }
+  image = "jetbrains/teamcity-server"
+}
+
+data "aws_region" "current" {}
```
Comment (henrykie) [inline modules/teamcity/local.tf:9]: pre-commit is not running correctly for you. Newlines like this should be caught, and Terraform should handle formatting and a README generation.

File: modules/teamcity/versions.tf [hunk 5]
```diff
@@ -0,0 +1,10 @@
+terraform {
+  required_version = ">= 1.0"
```
Comment (henrykie) [inline modules/teamcity/versions.tf:2]: also bump to 1.9.

File: modules/teamcity/versions.tf [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "5.82.2"
```
Comment (henrykie) [inline modules/teamcity/versions.tf:7]: Also bump to 5.85.0

File: modules/teamcity/outputs.tf [hunk 7]
```diff
@@ -0,0 +1,7 @@
+output "external_alb_dns_name" {
+  value = aws_lb.teamcity_external_lb.dns_name
```
Comment (henrykie) [inline modules/teamcity/outputs.tf:2]: add a description to this output

File: modules/teamcity/outputs.tf [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+}
+
+output "external_alb_zone_id" {
+  value = aws_lb.teamcity_external_lb.zone_id
```
Comment (henrykie) [inline modules/teamcity/outputs.tf:6]: add a description to this output

**Review by henrykie — CHANGES_REQUESTED** [review]
Great progress. Couple of minor things.

File: modules/teamcity/main.tf [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+  vpc_id      = var.vpc_id
+  target_type = "ip"
+
+  # health_check {
+  #   path                = "/"
+  #   interval            = 30
+  #   timeout             = 5
+  #   healthy_threshold   = 2
+  #   unhealthy_threshold = 2
+  # }
```
Comment (henrykie) [inline modules/teamcity/main.tf:286]: @cwwalb this is where we want to make adjustments to the healthcheck for TeamCity. Let's keep up our conversation here.

Comment (henrykie) [reply modules/teamcity/main.tf:286]: I think `/login` should be used instead of the root. The interval at 30 seconds and the unhealthy threshold should give TC enough time to boot to completion.

Comment (henrykie) [reply modules/teamcity/main.tf:286]: Wait! [TeamCity documentation points at `/healthCheck/healthy` and `/healthCheck/ready`](https://www.jetbrains.com/help/teamcity/teamcity-monitoring-and-diagnostics.html)

Let's use these :)

Comment (cwwalb) [reply modules/teamcity/main.tf:286]: Do you want 2 different health checks? One for if the server is still in configuration mode and one when its ready for requests? referring to the /healthy vs /ready

File: modules/teamcity/local.tf (hunk 4 above)
Comment (cwwalb) [reply modules/teamcity/local.tf:9]: How do I fix that?

File: modules/teamcity/main.tf (hunk 9 above)
Comment (henrykie) [reply modules/teamcity/main.tf]: I think it makes sense to pick just one. At your discretion / testing to determine which one you want to use.

File: modules/teamcity/main.tf [hunk 10]
```diff
@@ -0,0 +1,492 @@
...
+    aws_security_group.teamcity_db_sg.id
+  ]
+
+
+  serverlessv2_scaling_configuration {
+    max_capacity             = 1.0
+    min_capacity             = 0.0
+    seconds_until_auto_pause = 3600
+  }
+}
```
[the remark this answers is not in the document]
Comment (henrykie) [reply modules/teamcity/main.tf:384]: @cwwalb This is what I was referring to. We likely want to use IAM based authentication for connections from service -> cluster. I am not sure this works with the basic driver, so we probably need to do some more research here. Like, are the DB username and password environment variables we were passing optional? I am not sure.

File: modules/teamcity/local.tf [hunk 11]
```diff
@@ -0,0 +1,11 @@
+locals {
+  name_prefix = "teamcity"
+  tags = merge(var.tags, {
+    "environment" = var.environment
+  })
+  image                      = "jetbrains/teamcity-server"
+  database_connection_string = "jdbc:postgresql://teamcity-cluster.cluster-cwmktevrnnqg.us-east-1.rds.amazonaws.com:5432/teamcity"
+  database_user              = "teamcity"
+  database_password          = "teamcity2025"
```
Comment (henrykie) [inline modules/teamcity/local.tf:9]: I am not sure how [IAM database authentication](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/UsingWithRDS.IAMDBAuth.html) works, but I'll do some digging here to determine if this is necessary. These should certainly be variables configured by the user (with defaults) if we cannot use IAM based authentication.

File: modules/teamcity/main.tf [hunk 12]
```diff
@@ -0,0 +1,467 @@
...
+
+# ECS Service
+resource "aws_ecs_service" "teamcity" {
+  name                   = local.name_prefix
+  cluster                = aws_ecs_cluster.teamcity_cluster.name
+  task_definition        = aws_ecs_task_definition.teamcity_task_definition.arn
+  launch_type            = "FARGATE"
+  desired_count          = 1    #TODO: make this configurable
+  force_new_deployment   = true #TODO: make this configurable
+  enable_execute_command = true #TODO: make this configurable
```
Comment (henrykie) [inline modules/teamcity/main.tf:119]: These should depend on a `debug` variable.

File: modules/teamcity/main.tf [hunk 13]
```diff
@@ -0,0 +1,467 @@
...
+      "ssmmessages:OpenDataChannel",
+      "ssmmessages:OpenControlChannel",
+      "ssmmessages:CreateDataChannel",
+      "ssmmessages:CreateControlChannel",
+    ]
+    resources = [
+      "*"
+    ]
+  }
+}
```
Comment (henrykie) [inline modules/teamcity/main.tf:273]: I think we are probably missing some [EFS permissions](https://docs.aws.amazon.com/efs/latest/ug/iam-access-control-nfs-efs.html) - that could be impacting our access point usage.

**Review by henrykie — CHANGES_REQUESTED** [review]
Still a couple minor adjustments to make. We're getting closer though!

File: modules/teamcity/main.tf (hunk 9 above)
Comment (cwwalb) [reply modules/teamcity/main.tf]: that was fixed

File: modules/teamcity/main.tf (hunk 13 above)
Comment (henrykie) [reply modules/teamcity/main.tf:273]: @cwwalb Any further testing on this? Should we just save this for the access point issue you created?

File: modules/teamcity/local.tf (hunk 11 above)
Comment (henrykie) [reply modules/teamcity/local.tf:9]: @cwwalb any update on this?

File: modules/teamcity/main.tf (hunk 12 above)
Comment (henrykie) [reply modules/teamcity/main.tf:119]: @cwwalb any update on this?

File: modules/teamcity/local.tf (hunk 11 above)
Comment (cwwalb) [reply modules/teamcity/local.tf:9]: Looking into this today

File: modules/teamcity/main.tf (hunk 13 above)
Comment (cwwalb) [reply modules/teamcity/main.tf:273]: I don't think its iam because I tried elasticfilesystem:* on the file system and same error

Comment (cwwalb) [reply modules/teamcity/main.tf:273]: teamcity datadir is under /data/teamcity_server/datadir with the way we are deploying the path we are assigning to the tcuser is /services. Not sure why it is doing that

**Comment by kylesomers** [comment]
Review this - https://github.com/JetBrains/teamcity-amazon-ecs-plugin

Worth incorporating/extending?

**Comment by henrykie** [comment]
> Review this - https://github.com/JetBrains/teamcity-amazon-ecs-plugin
> 
> Worth incorporating/extending?

Yes - 100%.

File: modules/teamcity/variables.tf [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+
+variable "alb_subnets" {
+  type        = list(string)
+  description = "The subnets in which the ALB will be deployed"
+}
+
+variable "alb_certificate_arn" {
+  type        = string
+  description = "The ARN of the SSL certificate to use for the ALB"
+}
```
Comment (henrykie) [inline modules/teamcity/variables.tf:137]: These should be required **if** `create_external_alb = true` - otherwise they aren't required.

[See the docs for deets.](https://developer.hashicorp.com/terraform/language/expressions/custom-conditions#input-variable-validation)

**Review by henrykie — CHANGES_REQUESTED** [review]
Couple of small things. Looks good! Go ahead and mark this as "ready" so that it isn't a draft anymore.

File: modules/teamcity/examples/simple/main.tf [hunk 15]
```diff
@@ -0,0 +1,8 @@
+module "teamcity" {
+  source              = "../../"
+  vpc_id              = aws_vpc.teamcity_vpc.id
+  service_subnets     = aws_subnet.private_subnets[*].id
+  alb_subnets         = aws_subnet.public_subnets[*].id
+  alb_certificate_arn = aws_acm_certificate.teamcity.arn
+  debug               = true
```
[the remark this answers is not in the document]
Comment (cwwalb) [reply modules/teamcity/examples/simple/main.tf:7]: I forgot to delete that. I will delete I think its best kept in the documentation

File: modules/teamcity/main.tf [hunk 16]
```diff
@@ -0,0 +1,595 @@
...
+
+# File system for teamcity
+resource "aws_efs_file_system" "teamcity_efs_file_system" {
+  count            = var.efs_id != null ? 0 : 1
+  creation_token   = "${local.name_prefix}-efs-file-system"
+  performance_mode = var.teamcity_efs_performance_mode
+  throughput_mode  = var.teamcity_efs_throughput_mode
+
+  #TODO: Parameterize encryption and customer managed key creation
+  encrypted = false
```
[the remark this answers is not in the document]
Comment (cwwalb) [reply modules/teamcity/main.tf:466]: yes

File: modules/teamcity/main.tf [hunk 17]
```diff
@@ -0,0 +1,595 @@
...
+    actions = [
+      "elasticfilesystem:ClientWrite",
+      "elasticfilesystem:ClientRootAccess",
+      "elasticfilesystem:ClientMount"
+    ]
+    resources = [
+      local.efs_file_system_arn
+    ]
+  }
+}
```
[the remark this answers is not in the document]
Comment (cwwalb) [reply modules/teamcity/main.tf:287]: yes it does

File: modules/teamcity/main.tf [hunk 18]
```diff
@@ -0,0 +1,595 @@
...
+# Ingress rule for PostgreSQL from database cluster to service
+resource "aws_vpc_security_group_ingress_rule" "db_service" {
+  count                        = var.database_connection_string == null ? 1 : 0
+  security_group_id            = aws_security_group.teamcity_service_sg.id
+  referenced_security_group_id = aws_security_group.teamcity_db_sg[0].id
+  description                  = "Allow inbound access from DB to TeamCity service containers"
+  ip_protocol                  = "TCP"
+  from_port                    = 5432
+  to_port                      = 5432
+}
```
[the remark this answers is not in the document]
Comment (cwwalb) [reply modules/teamcity/main.tf:192]: you're right its not

File: modules/teamcity/main.tf (hunk 16 above)
[the remark this answers is not in the document]
Comment (cwwalb) [reply modules/teamcity/main.tf:461]: Actually scratch that its working and parameterized

File: modules/teamcity/examples/simple/local.tf [hunk 19]
```diff
@@ -0,0 +1,10 @@
+data "aws_availability_zones" "available" {}
+
+locals {
```
Comment (novekm) [inline modules/teamcity/examples/simple/local.tf:3]: We may want to consider abstracting this away to a data source or something in parameter store eventually. Ideally when we integrate with Terraform Test for automated tests, we should always reference the same VPC that will be in whichever AWS account we use for pipeline tests (values stored in SSM Parameter store or something). 

This will also help ensure customers don't run into issue if deploying quickly using two examples (e.g. TeamCity and Perforce) with overlapping IP address space. I think both examples currently use the same range.

File: modules/teamcity/examples/simple/main.tf [hunk 20]
```diff
@@ -0,0 +1,7 @@
+module "teamcity" {
+  source              = "../../"
```
Comment (novekm) [inline modules/teamcity/examples/simple/main.tf:2]: question for later - as we scale do we want to keep all of these TF modules local to the context of the repo, or do we want to start hosting these in the Terraform Registry and referencing that source? E.g. `source = aws-games/teamcity/aws`

File: modules/teamcity/examples/simple/versions.tf (hunk 2 above)
Comment (novekm) [reply modules/teamcity/examples/simple/versions.tf:2]: We should determine how/if we want to globally standardize and set these vales (Terraform version, AWS/AWSCC Provider version, etc.) and validate for all modules in the toolkit. Is there something in `v1.9` that may cause a breaking change and that's why you want to test that? Related, what is the oldest TF version we want to support? Open to discussion.

File: modules/teamcity/examples/simple/vpc.tf [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  tags = merge(local.tags,
+    {
+      Name = "teamcity-public-rt"
+    }
+  )
+}
+
+resource "aws_route_table_association" "public_rt_asso" {
```
Comment (novekm) [inline modules/teamcity/examples/simple/vpc.tf:76]: the "o" saved this resource naming 😂

File: modules/teamcity/examples/simple/vpc.tf [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+  route_table_id = aws_route_table.public_rt.id
+  subnet_id      = aws_subnet.public_subnets[count.index].id
+}
+
+resource "aws_eip" "nat_gateway_eip" {
+  depends_on = [aws_internet_gateway.igw]
+  #checkov:skip=CKV2_AWS_19:EIP associated with NAT Gateway through association ID
+  tags = merge(local.tags,
+    {
+      Name = "teamcity-nat-eip"
```
Comment (novekm) [inline modules/teamcity/examples/simple/vpc.tf:87]: I wonder if we should by default create a tag (at the module level) that also is like `ModuleSource = (path to module either in TF registry or in toolkit)` - could help with ease of navigation as we scale to more modules over time

File: modules/teamcity/examples/simple/local.tf (hunk 19 above)
Comment (henrykie) [reply modules/teamcity/examples/simple/local.tf:3]: WHat do you mean? For examples, we likely want to showcase VPC resource creation. Otherwise we're placing burden on end user to pre-create VPC.

If an end user wants to use a single VPC and deploy multiple modules, they should modify or augment one of our examples to suit their needs. Or just use the Toolkit as a submodule to pull our modules into existing TF.

File: modules/teamcity/examples/simple/main.tf (hunk 20 above)
Comment (henrykie) [reply modules/teamcity/examples/simple/main.tf:2]: Split this out into a discussion. Resolving the comment here.

File: modules/teamcity/examples/simple/vpc.tf (hunk 22 above)
Comment (henrykie) [reply modules/teamcity/examples/simple/vpc.tf:87]: Also good idea. Maybe propose in issue or in discussion. Lets keep these sorts of ideas out of the PR though to simplify review for Christina :)

File: modules/teamcity/main.tf
Comment (henrykie) [inline modules/teamcity/main.tf:1]: @cwwalb I would create an issue to track the work of splitting this file into separate files for readability.

**Review by henrykie — DISMISSED** [review]

**Comment by henrykie** [comment]
@cwwalb This looks great. Get any final changes you need in, rebase main to get up to date, and then re-request a review from me and I'll re-approve assuming everything deploys. @novekm is going to use this module as a guinea pig for Terraform test framework, so he'll give the other approval. Kevon, can you please check that everything deploys before we merge this as well? Just the single example Christina has provided - we can worry about edge cases and parameterized deployments (no ALB, existing DocDB, etc...) after. Thanks all!

**Review by novekm — COMMENTED** [review]
LGTM besides some minor comments - have you validated successful run of `terraform apply` and `terraform destroy` ?

File: modules/teamcity/local.tf [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+  efs_file_system_id  = var.efs_id != null ? var.efs_id : aws_efs_file_system.teamcity_efs_file_system[0].id
+  efs_file_system_arn = var.efs_id != null ? data.aws_efs_file_system.efs_file_system[0].arn : aws_efs_file_system.teamcity_efs_file_system[0].arn
+  efs_access_point_id = var.efs_access_point_id != null ? var.efs_access_point_id : aws_efs_access_point.teamcity_efs_data_access_point[0].id
+
+  # TeamCity Server Information
+  # Set environment variables
+  base_env = [
+    {
+      name  = "TEAMCITY_DB_URL"
+      value = local.database_connection_string
```
[the remark this answers is not in the document]
Comment (cwwalb) [reply modules/teamcity/local.tf:25]: locals.tf in the module

[the remark this answers is not in the document]
Comment (novekm) [reply modules/teamcity/local.tf:25]: Ah, I see it now - ty

**Review by novekm — APPROVED** [review]
LGTM

File: modules/teamcity/local.tf [hunk 24]
```diff
@@ -0,0 +1,50 @@
+locals {
+  name_prefix = "teamcity"
+  tags = merge(var.tags, {
+    "environment" = var.environment
+  })
+
+  # Database information
+  database_connection_string = var.database_connection_string != null ? var.database_connection_string : "jdbc:postgresql://${aws_rds_cluster.teamcity_db_cluster[0].endpoint}/teamcity"
+  database_master_username   = var.database_master_username != null ? var.database_master_username : aws_rds_cluster.teamcity_db_cluster[0].master_username
```
[the remark this answers is not in the document]
Comment (cwwalb) [reply modules/teamcity/local.tf:9]: great idea lets discuss this separately

**Review by henrykie — APPROVED** [review]
LGTM 🚀

**Review by novekm — APPROVED** [review]
LGTM 🚀

**Comment by awsbilal** [comment]
Terraform Destroy fails to delete resource "teamcity_alb_access_logs_bucket" due to items still in the bucket.
Adding "force_destroy = true" to the resource in main.tf seems to fix this issue.

resource "aws_s3_bucket" "teamcity_alb_access_logs_bucket" {
  count  = var.enable_teamcity_alb_access_logs && var.teamcity_alb_access_logs_bucket == null ? 1 : 0
  bucket = "${local.name_prefix}-alb-access-logs-${random_string.teamcity_alb_access_logs_bucket_suffix[0].result}"
  force_destroy = true
  tags = merge(local.tags, {
    Name = "${local.name_prefix}-alb-access-logs-${random_string.teamcity_alb_access_logs_bucket_suffix[0].result}"
  })
}
